File: sdk/typescript/src/ai/multimodal.ts
/**
 * Multimodal content helpers for AI prompts.
 * Provides Image, Audio, and File classes with factory methods for creating
 * multimodal content from various sources (files, URLs, buffers, base64).
 */

import { readFile } from 'node:fs/promises';
import { resolve } from 'node:path';

// MIME type mappings for common image formats
const IMAGE_MIME_TYPES: Record<string, string> = {
  '.jpg': 'image/jpeg',
  '.jpeg': 'image/jpeg',
  '.png': 'image/png',
  '.gif': 'image/gif',
  '.webp': 'image/webp',
  '.bmp': 'image/bmp',
};

// MIME type mappings for common audio formats
const AUDIO_MIME_TYPES: Record<string, string> = {
  '.wav': 'audio/wav',
  '.mp3': 'audio/mpeg',
  '.flac': 'audio/flac',
  '.ogg': 'audio/ogg',
};

const VIDEO_MIME_TYPES: Record<string, string> = {
  '.mp4': 'video/mp4',
  '.mpeg': 'video/mpeg',
  '.mpg': 'video/mpeg',
  '.mov': 'video/quicktime',
  '.webm': 'video/webm',
};

/**
 * Represents text content in a multimodal prompt.
 */
export class Text {
  readonly type: 'text' = 'text';
  readonly text: string;

  constructor(text: string) {
    this.text = text;
  }
}

/**
 * Represents image content in a multimodal prompt.
 */
export class Image {
  readonly type: 'image_url' = 'image_url';
  readonly imageUrl: { url: string; detail?: 'low' | 'high' | 'auto' };

  private constructor(imageUrl: { url: string; detail?: 'low' | 'high' | 'auto' }) {
    this.imageUrl = imageUrl;
  }

  /**
   * Create Image from a local file by converting to base64 data URL.
   */
  static async fromFile(
    filePath: string,
    detail: 'low' | 'high' | 'auto' = 'high'
  ): Promise<Image> {
    const absolutePath = resolve(filePath);

    // Read file and encode to base64
    const buffer = await readFile(absolutePath);
    const base64Data = buffer.toString('base64');

    // Determine MIME type from extension
    const ext = getExtension(absolutePath).toLowerCase();
    const mimeType = IMAGE_MIME_TYPES[ext] || 'image/jpeg';

    const dataUrl = `data:${mimeType};base64,${base64Data}`;
    return new Image({ url: dataUrl, detail });
  }

  /**
   * Create Image from a URL.
   */
  static fromUrl(
    url: string,
    detail: 'low' | 'high' | 'auto' = 'high'
  ): Image {
    return new Image({ url, detail });
  }

  /**
   * Create Image from a buffer.
   */
  static async fromBuffer(
    buffer: Buffer | Uint8Array,
    mimeType: string = 'image/jpeg',
    detail: 'low' | 'high' | 'auto' = 'high'
  ): Promise<Image> {
    const base64Data = Buffer.from(buffer).toString('base64');
    const dataUrl = `data:${mimeType};base64,${base64Data}`;
    return new Image({ url: dataUrl, detail });
  }

  /**
   * Create Image from a base64 string.
   */
  static async fromBase64(
    base64Data: string,
    mimeType: string = 'image/jpeg',
    detail: 'low' | 'high' | 'auto' = 'high'
  ): Promise<Image> {
    const dataUrl = `data:${mimeType};base64,${base64Data}`;
    return new Image({ url: dataUrl, detail });
  }
}

/**
 * Represents audio content in a multimodal prompt.
 */
export class Audio {
  readonly type: 'input_audio' = 'input_audio';
  readonly audio: { data: string; format: string };

  private constructor(audio: { data: string; format: string }) {
    this.audio = audio;
  }

  /**
   * Create Audio from a local file by converting to base64.
   */
  static async fromFile(
    filePath: string,
    format?: 'wav' | 'mp3' | 'flac' | 'ogg'
  ): Promise<Audio> {
    const absolutePath = resolve(filePath);

    // Auto-detect format from extension if not provided
    const ext = getExtension(absolutePath).toLowerCase().replace('.', '');
    const audioFormat =
      format ||
      (['wav', 'mp3', 'flac', 'ogg'].includes(ext) ? (ext as 'wav' | 'mp3' | 'flac' | 'ogg') : 'wav');

    // Read file and encode to base64
    const buffer = await readFile(absolutePath);
    const base64Data = buffer.toString('base64');

    return new Audio({ data: base64Data, format: audioFormat });
  }

  /**
   * Create Audio from a URL (downloads and converts to base64).
   */
  static async fromUrl(
    url: string,
    format: 'wav' | 'mp3' | 'flac' | 'ogg' = 'wav'
  ): Promise<Audio> {
    try {
      const response = await fetch(url);
      if (!response.ok) {
        throw new Error(`Failed to fetch audio from URL: ${response.status} ${response.statusText}`);
      }
      const arrayBuffer = await response.arrayBuffer();
      const base64Data = Buffer.from(arrayBuffer).toString('base64');
      return new Audio({ data: base64Data, format });
    } catch (error) {
      if (error instanceof TypeError && error.message.includes('fetch')) {
        throw new Error('URL download requires a fetch-compatible environment');
      }
      throw error;
    }
  }

  /**
   * Create Audio from a buffer.
   */
  static async fromBuffer(
    buffer: Buffer | Uint8Array,
    format: 'wav' | 'mp3' | 'flac' | 'ogg' = 'wav'
  ): Promise<Audio> {
    const base64Data = Buffer.from(buffer).toString('base64');
    return new Audio({ data: base64Data, format });
  }

  /**
   * Create Audio from a base64 string.
   */
  static async fromBase64(
    base64Data: string,
    format: 'wav' | 'mp3' | 'flac' | 'ogg' = 'wav'
  ): Promise<Audio> {
    return new Audio({ data: base64Data, format });
  }
}

/**
 * Represents video content in a multimodal prompt.
 */
export class Video {
  readonly type: 'video_url' = 'video_url';
  readonly videoUrl: { url: string };

  private constructor(videoUrl: { url: string }) {
    this.videoUrl = videoUrl;
  }

  /**
   * Create Video from a local file by converting to a base64 data URL.
   */
  static async fromFile(filePath: string): Promise<Video> {
    const absolutePath = resolve(filePath);
    const buffer = await readFile(absolutePath);
    const base64Data = buffer.toString('base64');
    const ext = getExtension(absolutePath).toLowerCase();
    const mimeType = VIDEO_MIME_TYPES[ext] || 'video/mp4';
    return new Video({ url: `data:${mimeType};base64,${base64Data}` });
  }

  /**
   * Create Video from a URL.
   */
  static fromUrl(url: string): Video {
    return new Video({ url });
  }

  /**
   * Create Video from a buffer.
   */
  static async fromBuffer(
    buffer: Buffer | Uint8Array,
    mimeType: string = 'video/mp4'
  ): Promise<Video> {
    const base64Data = Buffer.from(buffer).toString('base64');
    return new Video({ url: `data:${mimeType};base64,${base64Data}` });
  }

  /**
   * Create Video from a base64 string.
   */
  static async fromBase64(
    base64Data: string,
    mimeType: string = 'video/mp4'
  ): Promise<Video> {
    return new Video({ url: `data:${mimeType};base64,${base64Data}` });
  }
}

/**
 * Represents a generic file content in a multimodal prompt.
 */
export class File {
  readonly type: 'file' = 'file';
  readonly file: { url: string; mimeType?: string };

  private constructor(file: { url: string; mimeType?: string }) {
    this.file = file;
  }

  /**
   * Create File from a local file path.
   */
  static async fromFile(filePath: string, mimeType?: string): Promise<File> {
    const absolutePath = resolve(filePath);

    // Auto-detect MIME type if not provided
    const detectedMimeType =
      mimeType ||
      guessMimeType(absolutePath) ||
      'application/octet-stream';

    const buffer = await readFile(absolutePath);
    const base64Data = buffer.toString('base64');
    const dataUrl = `data:${detectedMimeType};base64,${base64Data}`;
    return new File({ url: dataUrl, mimeType: detectedMimeType });
  }

  /**
   * Create File from a URL.
   */
  static fromUrl(url: string, mimeType?: string): File {
    return new File({ url, mimeType });
  }

  /**
   * Create File from a buffer.
   */
  static async fromBuffer(buffer: Buffer | Uint8Array, mimeType: string): Promise<File> {
    const base64Data = Buffer.from(buffer).toString('base64');
    const dataUrl = `data:${mimeType};base64,${base64Data}`;
    return new File({ url: dataUrl, mimeType });
  }

  /**
   * Create File from a base64 string.
   */
  static async fromBase64(base64Data: string, mimeType: string): Promise<File> {
    const dataUrl = `data:${mimeType};base64,${base64Data}`;
    return new File({ url: dataUrl, mimeType });
  }
}

// Utility functions

/**
 * Get the file extension from a path.
 */
function getExtension(filePath: string): string {
  const lastDot = filePath.lastIndexOf('.');
  if (lastDot === -1) {
    return '';
  }
  return filePath.slice(lastDot);
}

/**
 * Guess MIME type from file extension.
 */
function guessMimeType(filePath: string): string | null {
  const ext = getExtension(filePath).toLowerCase();

  // Check image types
  if (ext in IMAGE_MIME_TYPES) {
    return IMAGE_MIME_TYPES[ext];
  }

  // Check audio types
  if (ext in AUDIO_MIME_TYPES) {
    return AUDIO_MIME_TYPES[ext];
  }

  if (ext in VIDEO_MIME_TYPES) {
    return VIDEO_MIME_TYPES[ext];
  }

  // Common document types
  const documentMimeTypes: Record<string, string> = {
    '.pdf': 'application/pdf',
    '.doc': 'application/msword',
    '.docx': 'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
    '.xls': 'application/vnd.ms-excel',
    '.xlsx': 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
    '.txt': 'text/plain',
    '.csv': 'text/csv',
    '.html': 'text/html',
    '.json': 'application/json',
    '.xml': 'application/xml',
    '.zip': 'application/zip',
  };

  return documentMimeTypes[ext] || null;
}

// Convenience factory functions

/**
 * Create text content.
 */
export function text(content: string): Text {
  return new Text(content);
}

/**
 * Create image content from a local file.
 */
export async function imageFromFile(
  filePath: string,
  detail: 'low' | 'high' | 'auto' = 'high'
): Promise<Image> {
  return Image.fromFile(filePath, detail);
}

/**
 * Create image content from a URL.
 */
export function imageFromUrl(
  url: string,
  detail: 'low' | 'high' | 'auto' = 'high'
): Image {
  return Image.fromUrl(url, detail);
}

/**
 * Create image content from a buffer.
 */
export async function imageFromBuffer(
  buffer: Buffer | Uint8Array,
  mimeType: string = 'image/jpeg',
  detail: 'low' | 'high' | 'auto' = 'high'
): Promise<Image> {
  return Image.fromBuffer(buffer, mimeType, detail);
}

/**
 * Create image content from a base64 string.
 */
export async function imageFromBase64(
  base64Data: string,
  mimeType?: string,
  detail: 'low' | 'high' | 'auto' = 'high'
): Promise<Image> {
  return Image.fromBase64(base64Data, mimeType, detail);
}

/**
 * Create audio content from a local file.
 */
export async function audioFromFile(
  filePath: string,
  format?: 'wav' | 'mp3' | 'flac' | 'ogg'
): Promise<Audio> {
  return Audio.fromFile(filePath, format);
}

/**
 * Create audio content from a URL.
 */
export async function audioFromUrl(
  url: string,
  format: 'wav' | 'mp3' | 'flac' | 'ogg' = 'wav'
): Promise<Audio> {
  return Audio.fromUrl(url, format);
}

/**
 * Create audio content from a buffer.
 */
export async function audioFromBuffer(
  buffer: Buffer | Uint8Array,
  format: 'wav' | 'mp3' | 'flac' | 'ogg' = 'wav'
): Promise<Audio> {
  return Audio.fromBuffer(buffer, format);
}

/**
 * Create audio content from a base64 string.
 */
export async function audioFromBase64(
  base64Data: string,
  format: 'wav' | 'mp3' | 'flac' | 'ogg' = 'wav'
): Promise<Audio> {
  return Audio.fromBase64(base64Data, format);
}

/**
 * Create video content from a local file.
 */
export async function videoFromFile(filePath: string): Promise<Video> {
  return Video.fromFile(filePath);
}

/**
 * Create video content from a URL.
 */
export function videoFromUrl(url: string): Video {
  return Video.fromUrl(url);
}

/**
 * Create video content from a buffer.
 */
export async function videoFromBuffer(
  buffer: Buffer | Uint8Array,
  mimeType: string = 'video/mp4'
): Promise<Video> {
  return Video.fromBuffer(buffer, mimeType);
}

/**
 * Create video content from a base64 string.
 */
export async function videoFromBase64(
  base64Data: string,
  mimeType: string = 'video/mp4'
): Promise<Video> {
  return Video.fromBase64(base64Data, mimeType);
}

/**
 * Create file content from a local file.
 */
export async function fileFromPath(filePath: string, mimeType?: string): Promise<File> {
  return File.fromFile(filePath, mimeType);
}

/**
 * Create file content from a URL.
 */
export function fileFromUrl(url: string, mimeType?: string): File {
  return File.fromUrl(url, mimeType);
}

/**
 * Create file content from a buffer.
 */
export async function fileFromBuffer(buffer: Buffer | Uint8Array, mimeType: string): Promise<File> {
  return File.fromBuffer(buffer, mimeType);
}

/**
 * Create file content from a base64 string.
 */
export async function fileFromBase64(base64Data: string, mimeType: string): Promise<File> {
  return File.fromBase64(base64Data, mimeType);
}

// Type for all multimodal content types
export type MultimodalContent = Text | Image | Audio | Video | File;

File: sdk/python/agentfield/multimodal.py
import base64
from pathlib import Path
from typing import Literal, Optional, Union

from pydantic import BaseModel, Field


class Text(BaseModel):
    """Represents text content in a multimodal prompt."""

    type: Literal["text"] = "text"
    text: str = Field(..., description="The text content.")


class Image(BaseModel):
    """Represents image content in a multimodal prompt."""

    type: Literal["image_url"] = "image_url"
    image_url: Union[str, dict] = Field(
        ...,
        description="The URL of the image, or a dictionary with 'url' and optional 'detail' (e.g., {'url': 'https://example.com/image.jpg', 'detail': 'high'}).",
    )

    @classmethod
    def from_file(cls, file_path: Union[str, Path], detail: str = "high") -> "Image":
        """Create Image from local file by converting to base64 data URL."""
        file_path = Path(file_path)
        if not file_path.exists():
            raise FileNotFoundError(f"Image file not found: {file_path}")

        # Read and encode image
        with open(file_path, "rb") as f:
            image_data = base64.b64encode(f.read()).decode()

        # Determine MIME type from extension
        ext = file_path.suffix.lower()
        mime_types = {
            ".jpg": "image/jpeg",
            ".jpeg": "image/jpeg",
            ".png": "image/png",
            ".gif": "image/gif",
            ".webp": "image/webp",
            ".bmp": "image/bmp",
        }
        mime_type = mime_types.get(ext, "image/jpeg")

        data_url = f"data:{mime_type};base64,{image_data}"
        return cls(image_url={"url": data_url, "detail": detail})

    @classmethod
    def from_url(cls, url: str, detail: str = "high") -> "Image":
        """Create Image from URL."""
        return cls(image_url={"url": url, "detail": detail})


class Audio(BaseModel):
    """Represents audio content in a multimodal prompt."""

    type: Literal["input_audio"] = "input_audio"
    input_audio: dict = Field(
        ..., description="Audio input data with 'data' (base64) and 'format' fields."
    )

    @classmethod
    def from_file(
        cls, file_path: Union[str, Path], format: Optional[str] = None
    ) -> "Audio":
        """Create Audio from local file by converting to base64."""
        file_path = Path(file_path)
        if not file_path.exists():
            raise FileNotFoundError(f"Audio file not found: {file_path}")

        # Auto-detect format from extension if not provided
        if format is None:
            ext = file_path.suffix.lower().lstrip(".")
            format = ext if ext in ["wav", "mp3", "flac", "ogg"] else "wav"

        # Read and encode audio
        with open(file_path, "rb") as f:
            audio_data = base64.b64encode(f.read()).decode()

        return cls(input_audio={"data": audio_data, "format": format})

    @classmethod
    def from_url(cls, url: str, format: str = "wav") -> "Audio":
        """Create Audio from URL (downloads and converts to base64)."""
        try:
            import requests

            response = requests.get(url)
            response.raise_for_status()
            audio_data = base64.b64encode(response.content).decode()
            return cls(input_audio={"data": audio_data, "format": format})
        except ImportError:
            raise ImportError("URL download requires requests: pip install requests")


class Video(BaseModel):
    """Represents video content in a multimodal prompt."""

    type: Literal["video_url"] = "video_url"
    video_url: dict = Field(
        ...,
        description="Video input data with a 'url' field containing an http(s) URL or data URL.",
    )

    @classmethod
    def from_file(cls, file_path: Union[str, Path]) -> "Video":
        """Create Video from local file by converting to a base64 data URL."""
        file_path = Path(file_path)
        if not file_path.exists():
            raise FileNotFoundError(f"Video file not found: {file_path}")

        with open(file_path, "rb") as f:
            video_data = base64.b64encode(f.read()).decode()

        ext = file_path.suffix.lower()
        mime_types = {
            ".mp4": "video/mp4",
            ".mpeg": "video/mpeg",
            ".mpg": "video/mpeg",
            ".mov": "video/quicktime",
            ".webm": "video/webm",
        }
        mime_type = mime_types.get(ext, "video/mp4")
        return cls(video_url={"url": f"data:{mime_type};base64,{video_data}"})

    @classmethod
    def from_url(cls, url: str) -> "Video":
        """Create Video from URL."""
        return cls(video_url={"url": url})


class File(BaseModel):
    """Represents a generic file content in a multimodal prompt."""

    type: Literal["file"] = "file"
    file: Union[str, dict] = Field(
        ...,
        description="The URL of the file, or a dictionary with 'url' and optional 'mime_type'.",
    )

    @classmethod
    def from_file(
        cls, file_path: Union[str, Path], mime_type: Optional[str] = None
    ) -> "File":
        """Create File from local file."""
        file_path = Path(file_path)
        if not file_path.exists():
            raise FileNotFoundError(f"File not found: {file_path}")

        # Auto-detect MIME type if not provided
        if mime_type is None:
            import mimetypes

            mime_type, _ = mimetypes.guess_type(str(file_path))
            mime_type = mime_type or "application/octet-stream"

        # For now, just store the file path - could be enhanced to base64 encode
        return cls(
            file={"url": f"file://{file_path.absolute()}", "mime_type": mime_type}
        )

    @classmethod
    def from_url(cls, url: str, mime_type: Optional[str] = None) -> "File":
        """Create File from URL."""
        return cls(file={"url": url, "mime_type": mime_type})


# Union type for all multimodal content types
MultimodalContent = Union[Text, Image, Audio, Video, File]


# Convenience functions for creating multimodal content
def text(content: str) -> Text:
    """Create text content."""
    return Text(text=content)


def image_from_file(file_path: Union[str, Path], detail: str = "high") -> Image:
    """Create image content from local file."""
    return Image.from_file(file_path, detail)


def image_from_url(url: str, detail: str = "high") -> Image:
    """Create image content from URL."""
    return Image.from_url(url, detail)


def audio_from_file(file_path: Union[str, Path], format: Optional[str] = None) -> Audio:
    """Create audio content from local file."""
    return Audio.from_file(file_path, format)


def audio_from_url(url: str, format: str = "wav") -> Audio:
    """Create audio content from URL."""
    return Audio.from_url(url, format)


def video_from_file(file_path: Union[str, Path]) -> Video:
    """Create video content from local file."""
    return Video.from_file(file_path)


def video_from_url(url: str) -> Video:
    """Create video content from URL."""
    return Video.from_url(url)


def file_from_path(
    file_path: Union[str, Path], mime_type: Optional[str] = None
) -> File:
    """Create file content from local file."""
    return File.from_file(file_path, mime_type)


def file_from_url(url: str, mime_type: Optional[str] = None) -> File:
    """Create file content from URL."""
    return File.from_url(url, mime_type)
